cpp/src/cylon/table.cpp

```diff
@@ -427,16 +497,298 @@ Status Sort(const std::shared_ptr<Table> &table, const std::vector<int32_t> &sor
+  }
+
+  CYLON_ASSIGN_OR_RAISE(auto take_arr, filter.Finish());
+  CYLON_ASSIGN_OR_RAISE(auto take_res, (arrow::compute::Take(local_sorted->get_table(), take_arr)));
+  sample_result = std::make_shared<Table>(ctx, take_res.table());
+
+  return Status::OK();
 }
 
-Status DistributedSort(const std::shared_ptr<Table> &table,
+template<typename T>
+static int CompareRows(const std::vector<std::unique_ptr<T>>& comparators,
+                      int64_t idx_a, int64_t idx_b) { // TODO: priority & direction
+  int sz = comparators.size();
+  if(std::is_same<T, cylon::DualArrayIndexComparator>::value) {
+    idx_b |= (int64_t)1 << 63;
+  }
+  for(int i = 0; i < sz; i++) {
+    int result = comparators[i]->compare(idx_a, idx_b);
+    if(result == 0) continue;
+    return result;
+  }
+  return 0;
+}
+
+Status MergeSortedTable(const std::vector<std::shared_ptr<Table>>& tables,
+                        const std::vector<int>& sort_columns,
+                        const std::vector<bool>& sort_orders,
+                        std::shared_ptr<Table>& out) {
+  std::shared_ptr<Table> concatenated;
+  std::vector<int64_t> table_indices(tables.size()), table_end_indices(tables.size());
+  int acc = 0;
+  for(int i = 0; i < table_indices.size(); i++) {
+    table_indices[i] = acc;
+    acc += tables[i]->Rows();
+    table_end_indices[i] = acc;
+  }
+
+  RETURN_CYLON_STATUS_IF_FAILED(Merge(tables, concatenated));
+
+  std::vector<std::unique_ptr<cylon::ArrayIndexComparator>> comparators(sort_columns.size());
+  for(int64_t i = 0; i < sort_columns.size(); i++) {
+    std::unique_ptr<cylon::ArrayIndexComparator> comp;
+    CreateArrayIndexComparator(
+      cylon::util::GetChunkOrEmptyArray(concatenated->get_table()->column(sort_columns[i]), 0), &comp, sort_orders[i]
+    );
+    comparators[i] = std::move(comp);
+  }
+
+  auto comp = [&](int a, int b) { // a and b are index of table in `tables`
+    int64_t a_idx = table_indices[a], b_idx = table_indices[b];
+    bool temp = CompareRows(comparators, a, b) > 0;
+    return CompareRows(comparators, a_idx, b_idx) > 0;
+  };
+
+  std::priority_queue<int, std::vector<int>, decltype(comp)> pq(comp);
+
+  for(int i = 0; i < tables.size(); i++) {
+    if(table_indices[i] < table_end_indices[i]) {
+      pq.push(i);
+    }
+  }
+
+  auto ctx = concatenated->GetContext();
+  auto pool = cylon::ToArrowPool(ctx);
+  arrow::Int64Builder filter(pool);
+  RETURN_CYLON_STATUS_IF_ARROW_FAILED(filter.Reserve(concatenated->Rows()));
+
+  std::vector<int> temp_v;
+
+  while(!pq.empty()) {
+    int t = pq.top();
+    pq.pop();
+    // std::cout<<table_indices[t]<<std::endl;
+    temp_v.push_back(table_indices[t]);
+    filter.UnsafeAppend(table_indices[t]);
+    table_indices[t] += 1;
+    if(table_indices[t] < table_end_indices[t]) {
+      pq.push(t);
+    }
+  }
+  
+  CYLON_ASSIGN_OR_RAISE(auto take_arr, filter.Finish());
+  CYLON_ASSIGN_OR_RAISE(auto take_res, (arrow::compute::Take(concatenated->get_table(), take_arr)));
```

Comment:
The sub-arrays are not sorted in initial sampling, but they are sorted in regular sampling. As result, we can use a merge in regular sampling, which is n log(k) in time (n = num items, k = world size). In initial sampling, we need to sort, which is n log(n) in time.

Do you think using merge will benefit us (e.g. more performant)? Or is it actually slower because of more overhead?

Reply:
(we already discussed this offline)